TechnicalDocuments/ControlTheWorld.md

```diff
@@ -0,0 +1,120 @@
+
+## Motivation
+
+To overcome some of the disadvantages of dependency injection:
+
+* Adding a new dependency a long way down the chain can mean passing that dependency all the way down through many layers. Many of those layers only need to get passed the dependency in order to pass it down to the next layer. This leads to a lot of source files being changed just to add a single dependency.
+
+* Many of our dependencies, such as the Analytics service or the AB Testing Service are, in reality, singletons which never change at runtime and are just passed all over the place.
+
+* In order to gain the testing benefit of dependency injection (being able to substitute a dependency with a stub implementation for testing), you have to use Protocols, which adds unnecessary boilerplate.
+
+## Benefits
+
+* Avoid having to pass singleton dependencies down through multiple layers of injection.
+
+* Ability to easily create different app "environments" to simulate running on different dates, timezones and locales, for example.
+
+* Ability to have "mock" environments for testing. For example, testing that analytics events are correctly being sent is typically quite difficult. With a mock analytics service in a mock app environment this becomes trivial, and without the need to inject an analytics service protocol to every part of the app.
+
+## Implementation
+
+Define a `struct` to provide dependencies of the environment external to the app (the "World" in which the app lives), and static configuaration dependencies:
+
+```swift
+struct World {
+    var date = { Date () }
+    var calendar = Calendar.autoupdatingCurrent
+    var timezone = Timezone.autoupdatingCurrent
+    var locale = Locale.autoupdatingCurrent
+    var analytics = Analytics()
+}
+```
+
+We can make external services like `Analytics` mockable without needing protocols:
+
+```swift
+struct Analytics {
+    struct Event {
+        var name: String
+        var properties: [String: String]
+    }
+
+    var track = track(_:)
+}
+
+private func track(_ event: Analytics.Event) {
+    // track
+}
+
+extension Analytics {
+    static let mock = Analytics(track: { event in
+        print("Mock track", event)
+    })
+}
+```
+
+By pulling the implementation (the `track` function) out into a private free function, it can be substituted with a mock.
+
+## Creating the World
+
+Now all we need is to instantiate an instance of `World` called `Current` which is our current environment. By default this will instantiate the production dependencies:
+
+```swift
+var Current = World()
+```
+
+## Usage
+
+Everywhere where we are referring to one of the dependencies now stored in `Current`, use `Current`: eg, instead of `Date()` use `Current.date()`.
```

Comment:
> Everywhere where we are referring to one of the dependencies now stored in `Current`, use `Current`

That's a strangely worded sentence, imo. Perhaps

"Everywhere where we'd be using a dependency, use `Current` to access it instead. For example, instead of `Date()` use `Current.date()`."